### PR TITLE
yihan integrate team code into the Teams page

### PR DESCRIPTION
**Pull request by AriaYu927**
# Description
截屏2023-09-06 09 22 32

This is the 3rd PR of "Create editable 4-letter team code option" task.

## Related PRS (if any):
This frontend PR is related to the [#529](https://github.com/OneCommunityGlobal/HGNRest/pull/529) backend PR.

## Main changes explained:
- Update file mockAdminState.js, mockStates.js, Team.test.js and allTeamsReducer.test.js for updating mock and unit test.
- Update file allTeamsAction.js and allTeamsReducer.js for adding teamcode parameter to updateTeam function.
- Update file TeamTable.jsx for integrating team code input box into table.
- Update file Reports.jsx for passing hasPermission and updateTeam function to TeamTable component.
- Update file DeleteTeamPopup.jsx, Team.jsx, TeamStatusPopup.jsx, Teams.jsx

## How to test:
1. check into current branch on both frontend and backend.
2. do `npm run start:local` to run this PR locally
3. log as Owner user or other uses who has Edit team 4-digit code permission.
4. go to Reports → Reports → Teams page
5. verify if you can edit team code of any team on that page. And the code you input should be auto-saved successfully.

### CODE REVIEW

**Review by DavidC0126 — DISMISSED** [review]
Hi @AriaYu927 I tested your PR and it works as intended. I can edit any links as an owner, and also the 4-digit code will be auto-saved after editing.

https://github.com/OneCommunityGlobal/HighestGoodNetworkApp/assets/74082492/ab5cdc31-f3ad-4745-be8c-982ce81843a3

**Review by harshidad — COMMENTED** [review]
Hi @AriaYu927, To test this PR, I need owner account. But how can I create the owner user account ? As I tried to create new user using DevAdmin account. but I can't see the role as owner . Please see the attached screenshot.
Role

**Review by Haoj1 — DISMISSED** [review]
@AriaYu927 Great work. Team code can be edited and saved successfully in Teams page.

https://github.com/OneCommunityGlobal/HighestGoodNetworkApp/assets/99279909/8e4d3f8e-6a80-43ea-910b-9e6f3f82d8b3

**Review by robertoooc — CHANGES_REQUESTED** [review]
Hi @AriaYu927, I tested your branch and unfortunately wasn't able to confirm functionality for this pr. Whenever I changed the team code input it wouldn't save after a page refresh. I'm not sure if this might just be me or anyone else will also encounter this but below are the errors I received. I made sure I was using the correct backend branch, I also made sure I was in an owner account, I emptied cache and hard reload, and I even logged out while also clearing browser memory but still kept confronting this issue.

https://github.com/OneCommunityGlobal/HighestGoodNetworkApp/assets/117053202/c2f1ba90-0b01-44f8-9b57-b1c9357d3773





Screenshot 2023-09-06 at 4 44 22 PM

**Review by iTvX — DISMISSED** [review]
Hi @AriaYu927 ,I have checked your pr, it worked as described. Great job.


https://github.com/OneCommunityGlobal/HighestGoodNetworkApp/assets/46791134/41091530-501e-4481-950a-dff04990f142

File: src/components/Reports/TeamTable.jsx [hunk 1]
```diff
@@ -1,13 +1,82 @@
...
+    const [hasError, setHasError] = useState(false);
+    const fullCodeRegex = /^[A-Z]-[A-Z]{3}$/;
+
+    const handleOnChange = (value, team) => {
+      props.updateTeam(team.teamName, team._id, team.isActive, value);
+    };
+  
+    const handleCodeChange = e => {
+      let value = e.target.value;
+      console.log(value);
```
Comment (GaryB93) [inline src/components/Reports/TeamTable.jsx:23]: Advise removal of console.log

**Review by GaryB93 — APPROVED** [review]
The team code can now be edited from the Teams table on the Reports page. In testing, it also saves upon refreshing the page. I did notice a console.log and left a comment in the TeamTable.jsx file for it to be removed. Good work!

Screenshot 2023-09-08 at 1 54 54 PM

**Review by gabi-lima — DISMISSED** [review]
Well done! Working as it should.
Captura de tela 2023-09-10 020412
Captura de tela 2023-09-10 020404

**Review by winghojackyli — APPROVED** [review]
Hi @AriaYu927 , I've tested the PR and it worked great. 


https://github.com/OneCommunityGlobal/HighestGoodNetworkApp/assets/108389005/9d875502-5d6d-4406-8346-50b3ee1f7de1

File: src/components/Reports/TeamTable.jsx (hunk 1 above)
Comment (AriaYu927) [reply src/components/Reports/TeamTable.jsx]: Great catch! I missed this one when I deleted all of the console.log. I've removed this one and updated the PR, please check. Thank you!

**Comment by AriaYu927** [comment]
> Hi @AriaYu927, I tested your branch and unfortunately wasn't able to confirm functionality for this pr. Whenever I changed the team code input it wouldn't save after a page refresh. I'm not sure if this might just be me or anyone else will also encounter this but below are the errors I received. I made sure I was using the correct backend branch, I also made sure I was in an owner account, I emptied cache and hard reload, and I even logged out while also clearing browser memory but still kept confronting this issue.
> 
>  TeamCode.Vid.mov 
> Screenshot 2023-09-06 at 4 44 22 PM

Hi @robertoooc , thanks for your testing! I cannot reproduce the issue, and it seems like other people can also save the change successfully. I believe the problem you are facing is still related to backend. Did you run `npm run build` before `npm start`?

**Review by vikrambadhan — COMMENTED** [review]
Hi,  Great Job! I tested your PR and it works as intended!

https://github.com/OneCommunityGlobal/HighestGoodNetworkApp/assets/85417946/6b9f00b7-81ba-47b1-b7d1-3af0c7cd0b01

**Review by Shiwani99 — APPROVED** [review]
Hi @AriaYu927,

You have done a great job in integrating the 4 digit codes into team page. I tested your PR and it works as expected. I also tried checking if the changed 4 digit code is seen in other user profiles too and I was able to see the change in other user profiles as well. I also tried different inputs to check how it works. I'm attaching video and screenshots. 


https://github.com/OneCommunityGlobal/HighestGoodNetworkApp/assets/78906820/4281f642-4a61-4784-ab61-f2b14f9c6c4f

Screenshot 2023-09-14 at 2 04 58 AM

**Review by wantingxu7 — APPROVED** [review]
The PR works with BE#529. Good work!

FE#1266_while_editting
FE#1266_after_edditing

**Comment by Jiyuanxia** [comment]
Hi @AriaYu927 , I have reviewed your pull request and tested it. But the code is not saved after I refresh the page.
Edit code:
Screenshot 2023-09-21 224611
Refresh page:
Screenshot 2023-09-21 224624

**Comment by dah00** [comment]
Hey, PR works as intended. I can edit the code of any team and the code is auto-saved. Great job!
PR1266_correct

**Comment by AriaYu927** [comment]
> Hi @AriaYu927 , I have reviewed your pull request and tested it. But the code is not saved after I refresh the page. Edit code: Screenshot 2023-09-21 224611 Refresh page: Screenshot 2023-09-21 224624

Thanks for your testing, please check if you have checkout the correct backend branch and run `npm run build` before `npm start`.

**Review by harshidad — DISMISSED** [review]
I tested PR and it working fine!

https://github.com/OneCommunityGlobal/HighestGoodNetworkApp/assets/109878936/b91502fc-dd34-4fd5-bac7-b12c80333736

**Review by KurtisIvey — CHANGES_REQUESTED** [review]
No checks on the backend as to whether the user profile is an owner when trying to change the team code. 

I'm able to change team codes to whatever I want to via postman despite using the auth token of an Admin user. From my understanding, only an Owner role should be able to do this .In addition, If the digits for the team code are explicitly stated to be 4 digits, you should have some sort of validation for that. 

Would love to also see that sanitizing, validation, and escaping is performed on these inputs in the backend.

pr1266 no backend check for owner status when changing team code

**Comment by Lin1404** [comment]
Hi @AriaYu927 , I have tested your PR and it works as intended.

https://github.com/OneCommunityGlobal/HighestGoodNetworkApp/assets/78375725/a2c935b0-b943-48b6-97d6-7617a4ec861e

**Review by Lin1404 — APPROVED** [review]

**Review by StrawberryCalpico — APPROVED** [review]
Hi! I just reviewed your PR, it worked perfectly, good job


https://github.com/OneCommunityGlobal/HighestGoodNetworkApp/assets/114443664/d5164451-98fa-4d7b-a6bd-bf7049b4d360

**Review by malikjahanzaib — APPROVED** [review]
Hi @AriaYu927, I reviewed you PR locally and was able to change the team code using an owner account and changes persisted after refresh (autosaved). Nice job!!!

Test rec:


https://github.com/OneCommunityGlobal/HighestGoodNetworkApp/assets/46221334/fb63cde5-9cb2-4a16-a0e3-63938a3bf714

**Comment by AriaYu927** [comment]
> No checks on the backend as to whether the user profile is an owner when trying to change the team code.
> 
> I'm able to change team codes to whatever I want to via postman despite using the auth token of an Admin user. From my understanding, only an Owner role should be able to do this .In addition, If the digits for the team code are explicitly stated to be 4 digits, you should have some sort of validation for that.
> 
> Would love to also see that sanitizing, validation, and escaping is performed on these changing items in the backend since that's considered best practice. However, my approval for this isn't based on that since that's a larger backend issue to address overall.
> 
> Me changing the team code to whatever I want to via postman: pr1266 no backend check for owner status when changing team code

Hi @KurtisIvey , I've updated this PR, please check. Thank you!

**Comment by RheaWu1212** [comment]
Hi I tested your PR and it worked as expected! 


https://github.com/OneCommunityGlobal/HighestGoodNetworkApp/assets/109626741/3a799810-19bd-4a8c-99cc-49e5038c66bf

**Review by KurtisIvey — APPROVED** [review]
Fantastic job!

I'm no longer able to force a team code edit through postman, so the backend is now catching unauthorized user roles. Everything functions as intended. Admin and lower accounts cannot edit due to inactive input and my owner account can without issue.


Screenshot 2023-11-02 at 9 39 50 PM
Screenshot 2023-11-02 at 9 37 55 PM
Screenshot 2023-11-02 at 9 37 33 PM

**Pull request merged**